src/wazuh_modules/agent_upgrade/manager/wm_agent_upgrade_upgrades.c

```diff
@@ -0,0 +1,542 @@
+
+/**
+ * Send a close file command to an agent
+ * @param agent_id id of the agent
+ * @param wpk_file name of the file to close in the agent
+ * @return error code
+ * @retval OS_SUCCESS on success
+ * @retval OS_INVALID on errors
+ * */
+STATIC int wm_agent_upgrade_send_close(int agent_id, const char *wpk_file) __attribute__((nonnull));
+
+/**
+ * Send a sha1 command to an agent
+ * @param agent_id id of the agent
+ * @param wpk_file name of the file to calculate sha1 in the agent
+ * @param file_sha1 sha1 of the file in the manager to compare
+ * @return error code
+ * @retval OS_SUCCESS on success
+ * @retval OS_INVALID on errors
+ * */
+STATIC int wm_agent_upgrade_send_sha1(int agent_id, const char *wpk_file, const char *file_sha1) __attribute__((nonnull));
+
+/**
+ * Send an upgrade command to an agent
+ * @param agent_id id of the agent
+ * @param wpk_file name of the file with the installation files in the agent
+ * @param installer name of the installer to run in the agent
+ * @return error code
+ * @retval OS_SUCCESS on success
+ * @retval OS_INVALID on errors
+ * */
+STATIC int wm_agent_upgrade_send_upgrade(int agent_id, const char *wpk_file, const char *installer) __attribute__((nonnull));
+
+void wm_agent_upgrade_init_upgrade_queue() {
+    upgrade_queue = queue_init(MAX_AGENTS);
+}
+
+void wm_agent_upgrade_destroy_upgrade_queue() {
+    queue_free(upgrade_queue);
+}
+
+void wm_agent_upgrade_prepare_upgrades() {
+    unsigned int index = 0;
+    OSHashNode *node = NULL;
+
+    node = wm_agent_upgrade_get_first_node(&index);
+
+    while (node) {
+        int agent_key = atoi(node->key);
+        wm_agent_task *agent_task = (wm_agent_task *)node->data;
+
+        node = wm_agent_upgrade_get_next_node(&index, node);
+
+        if (!queue_full(upgrade_queue) && !queue_push_ex(upgrade_queue, agent_task)) {
+            wm_agent_upgrade_remove_entry(agent_key, 0);
+        } else {
+            mterror(WM_AGENT_UPGRADE_LOGTAG, WM_UPGRADE_UPGRADE_QUEUE_FULL, agent_key);
+            wm_agent_upgrade_remove_entry(agent_key, 1);
+        }
+    }
+}
+
+void* wm_agent_upgrade_dispatch_upgrades(void *arg) {
+    wm_manager_configs *config = (wm_manager_configs *)arg;
+    wm_upgrade_args *upgrade_config = NULL;
+
+    // Initialize threads count mutex
+    w_mutex_init(&upgrade_threads_mutex, NULL);
+
+    // Initialize threads count condition variable
+    w_cond_init(&upgrade_threads_cond, NULL);
+
+    while (1) {
+        w_mutex_lock(&upgrade_threads_mutex);
+
+        if (upgrade_threads_count < config->max_threads) {
+            w_mutex_unlock(&upgrade_threads_mutex);
+
+            wm_agent_task *agent_task = queue_pop_ex(upgrade_queue);
+
+            os_calloc(1, sizeof(wm_upgrade_args), upgrade_config);
+            upgrade_config->config = config;
+            upgrade_config->agent_task = agent_task;
+
+            w_mutex_lock(&upgrade_threads_mutex);
```

Comment:
This whole mutex locking-unlocking is very confusing and could be simplified a lot

```
while (1) {
    // Blocks until available thread is ready
    wm_agent_upgrade_wait_available_thread(config->max_threads);

    wm_agent_task *agent_task = queue_pop_ex(upgrade_queue);

    os_calloc(1, sizeof(wm_upgrade_args), upgrade_config);
    upgrade_config->config = config;
    upgrade_config->agent_task = agent_task;
    
    // Thread that will launch the upgrade
    w_create_thread(wm_agent_upgrade_start_upgrade, (void *)upgrade_config);
}
```

with:
```
void wm_agent_upgrade_wait_available_thread(int max_thread) {
    w_mutex_lock(&upgrade_threads_mutex);

    if(max_thread >= upgrade_threads_count) {
        // Wait until any thread finish its execution
        w_cond_wait(&upgrade_threads_cond, &upgrade_threads_mutex);
    }
    
    upgrade_threads_count++;

    w_mutex_unlock(&upgrade_threads_mutex);
}
```

Reply:
Done!